### PR TITLE
217 refactor data conversion pipeline

### LINKED ISSUE #217: [REFACTOR] Refactor dataset conversion pipeline to support runtime-selectable data schemas
### Motivation

During integration testing of the SHE inference branch, it became clear that `wf-psf` currently assumes a fixed dataset structure across training, evaluation, and inference workflows.

This assumption is too restrictive for downstream integration scenarios where different processing stages may provide different subsets of fields at runtime.

For example:

* **Inference workflows** may only provide:

  * `positions`
  * `seds`

* **Training / evaluation workflows** may additionally provide:

  * `sources`
  * `masks`

In upcoming WaveDiff releases, the inference module will also be reused for evaluation workflows, which requires the pipeline to support multiple dataset “contracts” depending on runtime context.

### Proposed Changes

Introduce a schema-driven dataset conversion and validation system allowing dataset requirements to be selected dynamically at runtime.

Key additions include:

* Dataset schema registry for runtime validation
* Support for multiple processing modes:

  * `TRAIN`
  * `EVALUATION`
  * `INFERENCE`
* Separation of:

  * dataset schema definitions
  * field conversion handlers
  * conversion context objects
* Runtime selection of schema mode through configuration
* Relaxation of hard-coded assumptions about required dataset fields
* Structured logging for conversion and validation operations

### New Inference Configuration Parameter

```yaml
schema_mode: INFERENCE
```

Supported modes:

* `INFERENCE`

  * Requires only `positions` and `seds`

* `EVALUATION`

  * Expects additional fields such as `sources` and optional masks

### Benefits

* Enables integration with external pipelines (e.g. SHE)
* Decouples dataset representation from workflow assumptions
* Improves flexibility for future instruments and runtime configurations
* Simplifies reuse of inference components for evaluation workflows
* Makes dataset validation explicit and mode-aware

### Validation

The refactor was validated by re-running:

* training
* evaluation
* inference
* mock SHE pipeline integration

with multiple dataset schemas. Results remained reproducible across workflows.

### LINKED ISSUE #218: [REFACTOR] Refactor SED processing using domain-specific conversion contexts
### Description of Refactoring/Improvement

Refactor the dataset conversion pipeline to decouple domain-specific processing logic (currently SED handling) from the generic TensorFlow dataset conversion workflow.

The current implementation of SED conversion requires additional runtime dependencies (`simPSF`, `n_bins_lambda`) which are specific to SED processing but were previously passed directly through the generic conversion pipeline.

This issue proposes introducing explicit conversion context objects and domain-scoped handlers to isolate field-specific processing requirements from the generic dataset conversion system.

---

### Goals and Objectives

* Decouple generic dataset conversion from SED-specific processing logic
* Introduce explicit runtime context objects for specialized handlers
* Improve extensibility for future instrument- or domain-specific processing
* Reduce hardcoded assumptions inside the conversion pipeline
* Improve readability and maintainability of handler execution flow

---

### Current Code Behavior

Previously, specialized handlers relied on generic converter method signatures that mixed common conversion behavior with domain-specific runtime arguments.

For example, SED processing required:

* `simPSF`
* `n_bins_lambda`

These parameters were threaded through generic conversion logic even though they were only relevant to SED handling.

This resulted in:

* tight coupling between conversion infrastructure and SED processing
* non-generic handler signatures
* reduced extensibility for future processing domains
* less clear separation of responsibilities

---

### Proposed Changes

Introduce:

* `ConversionContext`
* `SEDContext`
* handler registry-based dispatch

Example structure:

```python id="pw4b9f"
@dataclass(frozen=True)
class ConversionContext:
    seds: SEDContext | None = None


@dataclass(frozen=True)
class SEDContext:
    simPSF: Any
    n_bins_lambda: int
```

Handlers receive only the domain-specific context they require:

```python id="rpr4bx"
def process_seds_handler(converter, dataset, sed_context):
    return converter.process_seds(
        dataset,
        simPSF=sed_context.simPSF,
        n_bins_lambda=sed_context.n_bins_lambda,
    )
```

The converter dispatches handlers dynamically using schema-defined field handlers and matching context attributes.

---

### Expected Benefits

* Cleaner separation between generic and specialized conversion logic
* More maintainable handler interfaces
* Improved extensibility for future processing domains
* Easier integration of additional instrument-specific contexts
* Reduced coupling between dataset schemas and processing implementations
* Improved testability of specialized handlers

---

### Dependencies

Related components include:

* `DatasetSchema`
* schema handler registry
* `TensorFlowDatasetConverter`
* inference and evaluation pipelines
* SED generation utilities

Potential future extensions may include:

* `InstrumentContext`
* `PSFContext`
* mission-specific conversion handlers

---

### Testing Plan

Validation should include:

* unit tests for handler dispatch
* unit tests for `ConversionContext` resolution
* integration tests for SED processing using realistic `PSFSimulator` instances
* reproducibility checks for:

  * training
  * evaluation
  * inference workflows

---

### Additional Context

This refactor emerged during integration testing of external inference workflows (e.g. SHE pipeline integration), where different processing modes required more flexible runtime handling and clearer separation between dataset contracts and field processing logic.

The design intentionally aligns schema field names with conversion context attributes:

```python id="cfb2n7"
handlers["seds"]  <->  context.seds
```

to simplify dynamic dispatch and reduce converter-specific conditionals.

---

### Impact Assessment

This change affects core dataset conversion infrastructure but preserves external workflow behavior.

The refactor improves internal architecture while maintaining reproducibility across existing training, evaluation, and inference workflows.

---

### Next Steps

* Finalize handler registry abstraction
* Extend conversion context documentation
* Evaluate future instrument-specific context requirements
* Assess whether additional field domains should adopt dedicated contexts
* Continue reducing hardcoded assumptions in the conversion pipeline

### PR DESCRIPTION
**Pull request by jeipollack**
## Summary

Introduce schema-driven dataset conversion and domain-specific conversion contexts to support flexible runtime dataset contracts across training, evaluation, and inference workflows.

This PR refactors the dataset conversion pipeline to decouple dataset validation from field-specific processing logic, enabling workflows with different dataset requirements (e.g. SHE inference vs evaluation/training datasets).

Closes #217 
Closes #218 
---

## What’s changed

* Add schema-driven dataset conversion infrastructure
* Introduce runtime-selectable dataset modes:

  * `TRAIN`
  * `EVALUATION`
  * `INFERENCE`
* Add dataset schema registry and validation handling
* Introduce specialized conversion handlers module
* Add `ConversionContext` and `SEDContext` abstractions
* Refactor SED processing using domain-specific handler contexts
* Remove hardcoded required field assumptions from conversion workflows
* Refactor dataset constants to support schema-driven conversion and handler dispatch
* Add configurable `schema_mode` to inference configuration
* Improve logging throughout dataset conversion and inference flows
* Simplify configuration namespace handling
* Refactor TensorFlow conversion tests into schema-based contract tests
* Improve typing for lazily initialized adapters and internal pipeline state

---

## How to test / verify

* Run training workflow using standard training configuration
* Run evaluation workflow using `DatasetMode.EVALUATION`
* Run inference workflow using `DatasetMode.INFERENCE`
* Run inference workflow using reduced datasets containing only:

  * `positions`
  * `seds`
* Execute unit and integration tests:

```bash
pytest src/wf_psf/tests/
```

* Verify reproducibility of:

  * training
  * evaluation
  * inference outputs

---

## Scope

* [ ] Feature
* [ ] Bug fix
* [ ] Hotfix
* [ ] Documentation / process change
* [x] Internal / refactor
* [ ] Release

This PR is part of the broader effort to improve runtime flexibility and external pipeline interoperability for WaveDiff and downstream integrations (e.g. SHE workflows).

---

## Changelog

* [x] Changelog fragment added (if applicable)

---

## Reviewer Checklist

* [x] The PR targets the correct base branch (`develop`, or `main` for release PRs)
* [x] The PR is assigned to the developer
* [x] Appropriate labels are applied
* [x] The PR is included in relevant projects and/or milestones
* [x] Description clearly explains what has changed
* [x] Issue references included, if applicable
* [x] Code and documentation adhere to current standards (`ruff`)
* [x] Documentation updates included, if relevant
* [x] CI tests are passing
* [x] All reviewer comments have been addressed

---

## Next Steps / Notes (if applicable)

* Evaluate future extension of conversion contexts for additional instrument domains
* Continue reducing hardcoded assumptions in dataset preparation workflows
* Assess whether schema validation should eventually move earlier into dataset loading stages

### CODE REVIEW

**Comment by jeipollack** [comment]
I need to update the CI pipeline because of broken compatibility with python 3.9.

**Comment by jeipollack** [comment]
@ChaitanyaChawak I fixed the CI error. I dropped python 3.9 and python 3.12 from the matrix. 3.9 is too old for tensorflow 2.15 and 3.12 is too new. 

what remains is the changelog. I will work on it now.

**Comment by jeipollack** [comment]
hello @ChaitanyaChawak , I pushed the changelog. all is done on my side. :)

**Comment by jeipollack** [comment]
actually, I lied. I am checking the API documentation.

**Review by ChaitanyaChawak — APPROVED** [review]
